### PR TITLE
Text form field initial value

### PR DESCRIPTION
**Pull request by HansMuller**
Currently `TextFormField.initialValue` defaults to an empty string and is used to initialize the TextFormField controller's `TextEditingController.text`. If an explicit TextEditingController parameter is provided, it is counter-productive to unconditionally initialize it since doing so clears its state.

This PR changes the default value of `TextFormField.initialValue` to null and only uses it if the TextFormField's controller is internally generated. If a controller is specified, the `initialValue` is defined by the controller's text and the `initialValue` parameter must be null.

This is an incompatible change.

Without this change, it is difficult to build a lazily constructed list - like a `ListView` - of TextFormFields because each time a TextFormField is recreated its TextEditingController is effectively reset. Given this change the following app works as expected. The state of all 25 TextFormField's is preserved upon scrolling because their TextEditingControllers are persistent.

```dart
class FormFieldLifetime extends StatefulWidget {
  @override
  _FormFieldLifetimeState createState() => new _FormFieldLifetimeState();
}

class _FormFieldLifetimeState extends State<FormFieldLifetime> {
  final List<TextEditingController> _controllers = new List.generate(25, (int n) {
    return new TextEditingController();
  }).toList();

  @override
  Widget build(BuildContext context) {
    return new Scaffold(
      appBar: new AppBar(title: new Text("FormFieldLifetime")),
      body: new Form(
        child: new ListView(
          children: new List<Widget>.generate(_controllers.length, (int n) {
            return new TextFormField(
              controller: _controllers[n],
            );
          }).toList(),
        ),
      ),
    );
  }
}
```

### CODE REVIEW

File: packages/flutter/test/widgets/form_test.dart [hunk 1]
```diff
@@ -204,8 +204,8 @@ void main() {
     expect(editableText.widget.controller.text, equals('world'));
   });
 
-  testWidgets('Provide initial value to input when controller is specified', (WidgetTester tester) async {
```
Comment (gspencergoog) [inline packages/flutter/test/widgets/form_test.dart]: Don't you still want to *also* test that the initialValue gets set on the implicitly created controller?

Comment (HansMuller) [reply packages/flutter/test/widgets/form_test.dart]: Good point, yes.

**Review by gspencergoog — APPROVED** [review]

**Comment by gspencergoog** [comment]
32384589-a60f0e74-c078-11e7-9bc1-e5b5287aea9d

**Comment by acidjazz** [comment]
@HansMuller I'm having an issue trying out this solution [here](https://github.com/flutter/flutter/issues/11500#issuecomment-389280074)

**Pull request merged**